srcs/dfabuilder.hpp

```diff
@@ -0,0 +1,39 @@
+#ifndef _BUILDER_HPP__
+#define _BUILDER_HPP__
+#include <unordered_map>
+#include <unordered_set>
+#include "dfa.hpp"
+#include "state.hpp"
+namespace dfa{
+  class Dfa::DfaBuilder{
+      public:
+        virtual DfaBuilder & AddStates() = 0;
+        virtual DfaBuilder & SetTransitTable() = 0;
+        virtual DfaBuilder & SetAcceptStates() = 0;
+        virtual DfaBuilder & SetStartState() = 0;
+        virtual ~DfaBuilder()=default;
+        Dfa Build(){
+          Dfa dfa{};
+          dfa._acceptStates = _acceptStates;
+          dfa._startState = _startState;
+          dfa._states = _states;
+          dfa._table = _table;
+          Clean();
+          return dfa;
+        }
+        void Clean(){
+          _startState = -1;
+          _acceptStates.clear();
+          _states.clear();
+          _table.clear();
+        }
+      protected:
+        DfaBuilder() = default;
+        int  _startState;
+        std::unordered_set<int> _acceptStates;
```

Comment:
What is the value of the class other than adding a lot of boilerplate code?
I also dislike the fact that this class shares internal structures with Dfa.  This seems defeating encapsulation: if you ever want to change the internal representation of the DFA (say you discovered a better data structure than unordered_map), you need to change Dfa, Dfa::Builder, and all subclasses of Dfa::Builder, which would be a nightmare.

Reply:
Agreed with the argument that changing internal data structure of dfa will cause changes to the builder and all its subclasses by such design. In this case, I think the builder shouldn't be inherited by any subclasses. Whoever needs a Dfa can combine different different primitive operations provided by builder to create the Dfa it wants. This probably means the caller itself needs to have polymorphism.

Reply:
Refactored by applying the suggestion with slight tweak. Thoughts?

Reply:
I think it's much better.